content/blog/how-tikv-accesses-data.md

```diff
@@ -0,0 +1,263 @@
+
+
+### Multi Raft
+
+{{< figure
+    src="/img/blog/how-tikv-accesses-data/multiraft.png"
+    caption="Multi Raft"
+    number="" >}}
+
+Because one Raft Group only processes a limited amount of data, we split the data into multiple Raft Groups, each of which correspond to a Region. The way of splitting is to slice by range. That is, we sort the keys of the data in byte order to form an infinite sorted map, and then slice it into a segments of continuous key range. Each key range is treated as a Region. The range of the Region uses the front-end and back-open mode [start, end). The end key of the previous region is the start key of the next region.
+
+Regions in TiKV have a maximum size limit. When this threshold is exceeded, it splits into two Regions, such as [a, b)-> [a, ab) + [ab, b). Conversely, if there is very little data in the Region, it will be merged with the neighboring Regions to form a larger Region, such as `[a, ab) + [ab, b)-> [a, b)`.
+
+### Percolator
+
+For the same Region, we can guarantee the consistency of the key operations  through the Raft consensus protocol. But if we want to operate data that falls on different regions at the same time, we need distributed transactions to ensure the consistency of operations.
+
+The most common method of distributed transactions is two-phase commit, also known as 2PC. TiKV draws its inspirations of 2PC implementation from Google's Percolator, with some practical optimizations to provide the distributed transaction support. The underlying mechanism of Percolator is rather complicated. For illustration purposes, we highlight the following few points:
+
+First, Percolator needs a timestamp oracle (TSO) service to allocate global timestamps for transactions. This timestamp is monotonically increasing in time and globally unique. Any transaction gets a start timestamp (`startTS`) at the beginning, and then a commit timestamp (`commitTS`) when the transaction is committed.
+
+Percolator provides three column families (CF)—Lock, Data and Write. When a key-value pair is written, the lock of this key is placed in the Lock CF, and the actual value in the Data CF. If the write is successfully committed, the corresponding commit information will be stored in the Write CF.
+
+When the key is stored in the Data CF and the Write CF, the corresponding timestamp will be added to the Key — `startTS` for the Data CF, and `commitTS` for the Write CF.
+
+Suppose we need to write `a = 1`, first a `startTS`, 10 for example, is obtained from the TSO; then comes the PreWrite stage of Percolator, when data is written in the Lock CF and the Data CF, as shown below:
+
+```
+Lock CF: W a = lock
```

Comment:
Where do we explain what the `W`, `D`, `S`, and `R` here are?

Reply:
Good catch!